vnext/Shared/JSI/NapiJsiRuntime.cpp

```diff
@@ -0,0 +1,1992 @@
+  return m_runtime;
+}
+
+#pragma endregion HostFunctionWrapper
+
+#pragma region NapiPreparedJavaScript
+
+NapiJsiRuntime::NapiPreparedJavaScript::NapiPreparedJavaScript(
+    unique_ptr<const Buffer> serializedBuffer,
+    const shared_ptr<const Buffer> &sourceBuffer,
+    string sourceUrl)
+    : m_sourceBuffer{sourceBuffer},
+      m_serializedBuffer{std::move(serializedBuffer)},
+      m_sourceUrl{std::move(sourceUrl)} {}
+
+const Buffer &NapiJsiRuntime::NapiPreparedJavaScript::SerializedBuffer() const {
+  return *m_serializedBuffer;
+}
+
+const Buffer &NapiJsiRuntime::NapiPreparedJavaScript::SourceBuffer() const {
+  return *m_sourceBuffer;
+}
+
+const string &NapiJsiRuntime::NapiPreparedJavaScript::SourceUrl() const {
+  return m_sourceUrl;
+}
+
+#pragma endregion NapiPreparedJavaScript
+
+#pragma region VectorBuffer
+
+NapiJsiRuntime::VectorBuffer::VectorBuffer(vector<uint8_t> data) : m_data(std::move(data)) {}
+
+const uint8_t *NapiJsiRuntime::VectorBuffer::data() const {
+  return m_data.data();
+}
+
+size_t NapiJsiRuntime::VectorBuffer::size() const {
+  return m_data.size();
+}
+
+#pragma endregion VectorBuffer
+
+#pragma region error - handling
+
+// Throws facebook::jsi::JSError or facebook::jsi::JSINativeException from NAPI error.
+[[noreturn]] void NapiJsiRuntime::ThrowJsException(napi_status status) const {
+  napi_value jsError{};
+  CHECK_NAPI_ELSE_CRASH(napi_get_and_clear_last_exception(m_env, &jsError));
+
+  if (!m_pendingJSError && (status == napi_pending_exception || InstanceOf(jsError, m_value.Error))) {
+    AutoRestore<bool> setValue(const_cast<bool *>(&m_pendingJSError), true);
+    RewriteErrorMessage(jsError);
+
+    throw JSError(*const_cast<NapiJsiRuntime *>(this), ToJsiValue(jsError));
+  } else {
+    std::ostringstream errorString;
+    errorString << "A call to Chakra API returned error code 0x" << std::hex << status << '.';
+
+    throw JSINativeException(errorString.str().c_str());
+  }
+}
+
+// Throws facebook::jsi::JSINativeException with a message.
+[[noreturn]] void NapiJsiRuntime::ThrowNativeException(char const *errorMessage) const {
+  throw JSINativeException(errorMessage);
+}
+
+// Rewrites error messages to match the JSI unit test expectations.
+void NapiJsiRuntime::RewriteErrorMessage(napi_value jsError) const {
+  // The code below must work correctly even if the 'message' getter throws.
+  // In case when it throws, we ignore that exception.
+  napi_value message{};
+  napi_status status = napi_get_property(m_env, jsError, m_propertyId.message, &message);
+  if (status != napi_ok) {
+    // If the 'message' property getter throws, then we clear the exception and ignore it.
+    napi_value ignoreJSError{};
+    napi_get_and_clear_last_exception(m_env, &ignoreJSError);
+  } else if (TypeOf(message) == napi_string) {
+    // JSI unit tests expect V8- or JSC-like messages for the stack overflow.
+    if (StringToStdString(message) == "Out of stack space") {
```

Comment:
Can V8/NodeApi be localized. it seems 'risky' checking for an error substring here?

Reply:
If not the message below, should probably be localized by/for RN/RNW